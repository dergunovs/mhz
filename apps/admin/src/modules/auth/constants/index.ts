export const API_POST_LOGIN = 'auth/login';
export const API_GET_CHECK_AUTH = 'auth/check';

export const URL_LOGIN = '/';
