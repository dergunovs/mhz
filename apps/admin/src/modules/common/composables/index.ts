import { useRoute } from 'vue-router';
import { useQueryClient } from '@tanstack/vue-query';

const queryClient = useQueryClient(); // стоит проверить, вызвав invalidate

export function useInvalidate(queryKey: string) {
  const route = useRoute(); 
  function invalidate() {
    if (route.query.create === '1') {
      queryClient.invalidateQueries({ queryKey: [queryKey], exact: true });
    }
  }

  return {
    invalidate,
  };
}
