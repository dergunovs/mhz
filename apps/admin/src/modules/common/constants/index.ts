export const URL_MAIN = '/main';
export const URL_ERROR = '/404';
