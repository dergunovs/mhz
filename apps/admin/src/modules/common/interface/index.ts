import { ICategory, IManager, IManufacturer, IProduct } from 'mhz-types';

export interface ISearchResult {
  products: IProduct[];
  categories: ICategory[];
  manufacturers: IManufacturer[];
  managers: IManager[];
}
