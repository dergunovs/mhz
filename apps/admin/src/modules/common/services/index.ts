import { Ref } from 'vue';

import { api, useQuery, useMutation } from 'mhz-helpers';
import {
  API_UPLOAD,
  API_UPLOAD_SINGLE,
  API_SEARCH,
  API_STATS_COUNT,
  ISearchResults,
  IEntitiesCount,
  IBaseReply,
} from 'mhz-contracts';

export function search(query: Ref<string>, isAdmin?: boolean) {
  async function fn() {
    const { data } = await api.get<ISearchResults>(API_SEARCH, { params: { search: query.value, isAdmin } });

    return data;
  }

  return useQuery({ queryKey: [API_SEARCH, query], queryFn: fn, enabled: false });
}

export function getEntitiesCount() {
  async function fn() {
    const { data } = await api.get<IEntitiesCount>(API_STATS_COUNT);

    return data;
  }

  return useQuery({ queryKey: [API_STATS_COUNT], queryFn: fn });
}

export function uploadFile(options: object, width?: string) {
  async function fn(file?: File) {
    if (!file) throw new Error();

    const formData = new FormData();

    formData.append('file', file);

    const { data } = await api.post<string>(API_UPLOAD_SINGLE, formData, { params: { width } });

    return data;
  }

  return useMutation({ mutationKey: [API_UPLOAD], mutationFn: fn, ...options });
}

export function uploadFiles(options: object, width?: string, thumb?: boolean) {
  async function fn(files: File[]) {
    const formData = new FormData();

    files.forEach((file) => formData.append('files', file));

    const { data } = await api.post<string[]>(API_UPLOAD, formData, { params: { width, thumb } });

    return data;
  }

  return useMutation({ mutationKey: [API_UPLOAD], mutationFn: fn, ...options });
}

export function deleteFile(thumb?: boolean) {
  async function fn(filename: string) {
    const { data } = await api.delete<IBaseReply>(`${API_UPLOAD}/${filename}`, { params: { thumb } });

    return data;
  }

  return useMutation({ mutationKey: [API_UPLOAD], mutationFn: fn });
}
