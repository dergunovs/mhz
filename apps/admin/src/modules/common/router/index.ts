import { createRouter, createWebHistory } from 'vue-router';

import { routes } from '@/common/router/routes';
import { getCookieToken } from '@/auth/composables';
import { URL_LOGIN } from '@/auth/constants';

const router = createRouter({
  history: createWebHistory(),
  routes,
  scrollBehavior(to, from, savedPosition) {
    return savedPosition ? savedPosition : { top: 0 };
  },
});

router.beforeEach((to, from, next) => {
  if (to.path !== URL_LOGIN && !getCookieToken()) {
    next({ path: `${URL_LOGIN}?logout=1` });
  } else {
    next();
  }
});

export default router;
