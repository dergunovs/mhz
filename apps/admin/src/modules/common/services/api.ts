import axios from 'axios';

import { logout } from '@/auth/composables';

axios.defaults.baseURL = import.meta.env.VITE_API;

export const api = axios;

export function setAuthHeader(token: string) {
  api.defaults.headers.common['Authorization'] = `Bearer ${token}`;
}

export function deleteAuthHeader() {
  api.defaults.headers.common['Authorization'] = ``;
}

export function handleError(error: unknown) {
  if (axios.isAxiosError(error)) {
    if (error.response?.status === 403) logout();
    alert(error.response?.data.message);
  }
}
