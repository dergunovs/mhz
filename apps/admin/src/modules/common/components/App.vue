<template>
  <div>
    <component v-if="isLoaded" :is="layoutComponent" :data-layout="layoutComponent.name" data-test="app-layout" />
  </div>
</template>

<script setup lang="ts">
import { ref, computed, onMounted } from 'vue';
import { useRoute, useRouter } from 'vue-router';
import { useHead } from '@unhead/vue';

import { getCookieToken, setAuthHeader } from 'mhz-helpers';

import LayoutDefault from '@/layout/components/LayoutDefault.vue';
import LayoutEmpty from '@/layout/components/LayoutEmpty.vue';

import { checkAuth } from '@/auth/services';
import { TOKEN_NAME, URL_LOGIN } from '@/auth/constants';

const route = useRoute();
const router = useRouter();

useHead({
  meta: [{ name: 'version', content: import.meta.env.VITE_VERSION }],
});

const isLoaded = ref(false);

const layoutComponent = computed(() => {
  return route.meta.layout === 'empty' ? LayoutEmpty : LayoutDefault;
});

const isLoginPage = globalThis.location.pathname === URL_LOGIN;

const token = getCookieToken(TOKEN_NAME);

if (!isLoginPage && token) {
  setAuthHeader(token);

  checkAuth();
}

onMounted(async () => {
  await router.isReady();
  isLoaded.value = true;
});
</script>
