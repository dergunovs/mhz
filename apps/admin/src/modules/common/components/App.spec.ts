import { describe, it, expect, beforeEach, afterEach, vi, beforeAll } from 'vitest';
import { VueWrapper, enableAutoUnmount } from '@vue/test-utils';
import { IBaseReply } from 'mhz-contracts';
import * as helpers from 'mhz-helpers';

import App from './App.vue';

import { mockQueryReply, wrapperFactory, router } from '@/common/test';
import { TOKEN_NAME } from '@/auth/constants';
import { URL_CATEGORY } from '@/category/constants';

import * as authServices from '@/auth/services';

const spyGetCookieToken = vi.spyOn(helpers, 'getCookieToken').mockImplementation((token: string) => token);
const spySetAuthHeader = vi.spyOn(helpers, 'setAuthHeader');
const spyCheckAuth = vi.spyOn(authServices, 'checkAuth').mockReturnValue(mockQueryReply<IBaseReply>({ message: 'ok' }));

const layout = '[data-test="app-layout"]';

let wrapper: VueWrapper;

beforeAll(async () => {
  router.push('/');
  await router.isReady();
});

beforeEach(() => {
  wrapper = wrapperFactory(App, {});
});

enableAutoUnmount(afterEach);

describe('App', async () => {
  it('exists', async () => {
    expect(wrapper.findComponent(App)).toBeTruthy();
  });

  it('gets token from cookies and not sets auth at login page', async () => {
    expect(spyGetCookieToken).toBeCalledTimes(1);
    expect(spyGetCookieToken).toBeCalledWith(TOKEN_NAME);

    expect(spySetAuthHeader).toBeCalledTimes(0);
    expect(spyCheckAuth).toBeCalledTimes(0);
  });

  it('shows default layout and sets auth if token exists and page is not login', async () => {
    wrapper.unmount();

    Object.defineProperty(window, 'location', {
      value: new URL(`http://localhost:3000${URL_CATEGORY}`),
      writable: true,
    });

    const wrapperNotLogin = wrapperFactory(App, {});

    router.push(URL_CATEGORY);
    await router.isReady();

    await new Promise((r) => {
      setTimeout(r, 10);
    });

    expect(spySetAuthHeader).toBeCalledTimes(1);
    expect(spySetAuthHeader).toBeCalledWith(TOKEN_NAME);

    expect(spyCheckAuth).toBeCalledTimes(1);

    expect(wrapperNotLogin.find(layout).attributes('data-layout')).toBe('LayoutDefault');
  });
});
