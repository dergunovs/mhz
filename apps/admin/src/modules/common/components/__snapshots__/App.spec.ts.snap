// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`App > matches snapshot 1`] = `
"<div>
  <!--v-if-->
</div>"
`;
